packages/labs/ssr/src/test/lib/render-result-iterator_test.ts

```diff
@@ -0,0 +1,385 @@
+  const thunk: Thunk = () => 'thunk-value';
+  const thunked: ThunkedRenderResult = ['a', thunk, 'b'];
+  const iterator = new RenderResultIterator(thunked);
+
+  const result1 = iterator.next();
+  assert.equal(result1.done, false);
+  assert.equal(result1.value, 'a');
+
+  const result2 = iterator.next();
+  assert.equal(result2.done, false);
+  assert.equal(result2.value, 'thunk-value');
+
+  const result3 = iterator.next();
+  assert.equal(result3.done, false);
+  assert.equal(result3.value, 'b');
+
+  const result4 = iterator.next();
+  assert.equal(result4.done, true);
+});
+
+test('RenderResultIterator evaluates thunks returning arrays with nested thunks', () => {
+  const innerThunk: Thunk = () => 'inner-value';
+  const outerThunk: Thunk = () => [innerThunk];
+  const thunked: ThunkedRenderResult = ['a', outerThunk, 'b'];
+  const iterator = new RenderResultIterator(thunked);
+
+  const result1 = iterator.next();
+  assert.equal(result1.done, false);
+  assert.equal(result1.value, 'a');
+
+  const result2 = iterator.next();
+  assert.equal(result2.done, false);
+  assert.equal(result2.value, 'inner-value');
+
+  const result3 = iterator.next();
+  assert.equal(result3.done, false);
+  assert.equal(result3.value, 'b');
+});
+
+test('RenderResultIterator evaluates thunks returning arrays', () => {
+  const thunk: Thunk = () => ['x', 'y'];
+  const thunked: ThunkedRenderResult = ['a', thunk, 'b'];
+  const iterator = new RenderResultIterator(thunked);
+
+  const values: string[] = [];
+  let result = iterator.next();
+  while (!result.done) {
+    if (typeof result.value === 'string') {
+      values.push(result.value);
+    }
+    result = iterator.next();
+  }
+
+  assert.equal(values, ['a', 'x', 'y', 'b']);
+});
+
+test('RenderResultIterator evaluates thunks returning nested arrays', () => {
+  const innerThunk: Thunk = () => ['y', 'z'];
+  const outerThunk: Thunk = () => ['x', innerThunk];
+  const thunked: ThunkedRenderResult = ['a', outerThunk, 'b'];
+  const iterator = new RenderResultIterator(thunked);
+
+  const values: string[] = [];
+  let result = iterator.next();
+  while (!result.done) {
+    if (typeof result.value === 'string') {
+      values.push(result.value);
+    }
+    result = iterator.next();
+  }
+
+  assert.equal(values, ['a', 'x', 'y', 'z', 'b']);
+});
+
+//
+// Promise handling
+//
+
+test('RenderResultIterator converts Promise<string> to Promise<RenderResult>', async () => {
+  const thunk: Thunk = () => Promise.resolve('async-value');
+  const thunked: ThunkedRenderResult = ['a', thunk, 'b'];
+  const iterator = new RenderResultIterator(thunked);
+
+  const result1 = iterator.next();
+  assert.equal(result1.done, false);
+  assert.equal(result1.value, 'a');
+
+  const result2 = iterator.next();
+  assert.equal(result2.done, false);
+  assert.ok(result2.value instanceof Promise);
+
+  const resolvedValue = await result2.value;
+  assert.equal(resolvedValue, 'async-value');
```

Comment:
This relates to my earlier question about `this._waiting`. For completeness, should you also assert `'b'` or add a test with `for await` with this scenario?